votingsystem/src/voting.rs

```diff
@@ -0,0 +1,280 @@
+use coordinator::module::*;
+use coordinator::types::*;
+use foundry_module_rt::UserModule;
+use parking_lot::RwLock;
+use primitives::H256;
+use rand::Rng;
+use remote_trait_object::raw_exchange::{import_service_from_handle, HandleToExchange, Skeleton};
+use remote_trait_object::{Context as RtoContext, Service, ServiceRef};
+use serde::{Deserialize, Serialize};
+use staking::serialize;
+use std::sync::Arc;
+
+enum ExecuteError {
+    InvalidMetadata,
+    InvalidSign,
+    InvalidFormat,
+    InvalidAdmin,
+    UsedVotePaper,
+}
+
+pub enum Id {
+    VotePaperId(H256),
+    BallotId(u8),
+}
+#[derive(Debug, Serialize, Deserialize, Clone)]
+#[serde(transparent)]
+pub struct VoteId {
+    id: H256,
+}
+
+type StorageKeyRef = [u8];
+
+impl AsRef<StorageKeyRef> for VoteId {
+    fn as_ref(&self) -> &StorageKeyRef {
+        self.id.0.as_ref()
+    }
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+struct VotePaper {
+    vote_paper_id: H256,
+    general_meeting_id: GeneralMeetingId,
+    agenda_number: u32,
+    authorization_token: H256,
+    used_in: Option<VoteId>,
+    voter_publickey: Public,
+}
+
+impl VotePaper {
+    pub fn new(id: GeneralMeetingId, agenda_number: u32, token: H256, pub_key: Public) -> Self {
+        let mut rng = rand::thread_rng();
+        let random_id: u64 = rng.gen();
+        let vote_paper_id = H256::from(random_id);
+
+        Self {
+            vote_paper_id,
+            general_meeting_id: id,
+            agenda_number,
+            authorization_token: token,
+            used_in: None,
+            voter_publickey: pub_key,
+        }
+    }
+    pub fn set_used_in(&mut self, vote_id: VoteId) -> Result<(), ExecuteError> {
+        if self.used_in.is_none() {
+            return Err(ExecuteError::UsedVotePaper)
+        }
+        self.used_in = Some(vote_id);
+        Ok(())
+    }
+
+    pub fn get_general_meeting_id(&self) -> GeneralMeetingId {
+        self.general_meeting_id.clone()
+    }
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+struct Vote {
+    ballot_id: VoteId,
+    choice: VoteResult,
+    signature: Signature,
+}
+
+impl Vote {
+    pub fn new(choice: VoteResult, signature: Signature) -> Self {
+        let mut rng = rand::thread_rng();
+        let random_id: u64 = rng.gen();
+        let ballot_id = VoteId {
+            id: H256::from(random_id),
+        };
+        Self {
+            ballot_id,
+            choice,
+            signature,
+        }
+    }
+}
+struct Context {
+    pub storage: Option<Box<dyn SubStorageAccess>>,
+}
+
+impl Context {
+    fn storage(&self) -> &dyn SubStorageAccess {
+        self.storage.as_ref().unwrap().as_ref()
+    }
+
+    fn storage_mut(&mut self) -> &mut dyn SubStorageAccess {
+        self.storage.as_mut().unwrap().as_mut()
+    }
+}
+
+impl Service for Context {}
+
+#[derive(Serialize, Deserialize, Debug)]
+pub struct TxCreateVotePaper {
+    general_meeting_id: GeneralMeetingId,
+    agenda_number: u32,
+    authorization_token: H256,
```

Comment:
What is the `authorization_token`?

Reply:
It is a token that the person gets after authentication. We should check whether the person is authorized or not but because we do not save their tokens in the first place it needs a mechanism for it.

Reply:
I think my explanation was not good.

We can think about it like this:
Actually, creating a vote paper (with the public key) means that a "person" who has a signature is a stakeholder.
When a voter sends vote transaction, we can verify the vote by checking the signature is created by the "private key".